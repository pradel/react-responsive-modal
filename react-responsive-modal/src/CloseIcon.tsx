import React from 'react';
import cx from 'classnames';

interface CloseIconProps {
  id?: string;
  closeIcon?: React.ReactNode;
  styles?: {
    closeButton?: React.CSSProperties;
    closeIcon?: React.CSSProperties;
  };
  classNames?: {
    closeButton?: string;
    closeIcon?: string;
  };
  classes: {
    closeButton?: string;
  };
  onClick: () => void;
}

const CloseIcon = ({
  classes,
  classNames,
  styles,
  id,
  closeIcon,
  onClick,
}: CloseIconProps) => (
  <button
    id={id}
    className={cx(classes.closeButton, classNames?.closeButton)}
    style={styles?.closeButton}
    onClick={onClick}
    data-testid="close-button"
  >
    {closeIcon ? (
      closeIcon
    ) : (
      <svg
        className={classNames?.closeIcon}
        style={styles?.closeIcon}
        width={28}
        height={28}
        viewBox="0 0 36 36"
        data-testid="close-icon"
      >
        <path d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z" />
      </svg>
    )}
  </button>
);

export default CloseIcon;
