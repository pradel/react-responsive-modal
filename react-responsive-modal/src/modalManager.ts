import { Ref, useEffect } from 'react';

let modals: Ref<Element>[] = [];

/**
 * Handle the order of the modals.
 * Inspired by the material-ui implementation.
 */
export const modalManager = {
  /**
   * Register a new modal
   */
  add: (newModal: Ref<Element>) => {
    modals.push(newModal);
  },

  /**
   * Remove a modal
   */
  remove: (oldModal: Ref<Element>) => {
    modals = modals.filter((modal) => modal !== oldModal);
  },

  /**
   * When multiple modals are rendered will return true if current modal is the last one
   */
  isTopModal: (modal: Ref<Element>) =>
    !!modals.length && modals[modals.length - 1] === modal,
};

export function useModalManager(ref: Ref<Element>, open: boolean) {
  useEffect(() => {
    if (open) {
      modalManager.add(ref);
    }
    return () => {
      modalManager.remove(ref);
    };
  }, [open, ref]);
}
