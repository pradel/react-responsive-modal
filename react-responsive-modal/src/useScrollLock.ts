import { useEffect, useRef } from 'react';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

export const useScrollLock = (
  refModal: React.RefObject<Element>,
  open: boolean,
  showPortal: boolean,
  blockScroll: boolean,
  reserveScrollBarGap?: boolean
) => {
  const oldRef = useRef<Element | null>(null);

  useEffect(() => {
    if (open && refModal.current && blockScroll) {
      oldRef.current = refModal.current;
      disableBodyScroll(refModal.current, { reserveScrollBarGap });
    }
    return () => {
      if (oldRef.current) {
        enableBodyScroll(oldRef.current);
        oldRef.current = null;
      }
    };
  }, [open, showPortal, refModal, blockScroll, reserveScrollBarGap]);
};
