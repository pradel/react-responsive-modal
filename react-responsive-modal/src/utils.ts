import noScroll from 'no-scroll';

export const isBrowser = typeof window !== 'undefined';

export const blockNoScroll = () => {
  noScroll.on();
};

export const unblockNoScroll = () => {
  noScroll.off();
};
