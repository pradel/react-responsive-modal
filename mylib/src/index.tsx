import React, { useEffect, useState, useRef } from 'react';
import ReactDom from 'react-dom';
// import cx from 'classnames';
import CSSTransition from 'react-transition-group/CSSTransition';
import noScroll from 'no-scroll';
import modalManager from './modalManager';

const isBrowser = typeof window !== 'undefined';

const blockNoScroll = () => {
  noScroll.on();
};

const unblockNoScroll = () => {
  // Restore the scroll only if there is no modal on the screen
  if (modalManager.modals().length === 0) {
    noScroll.off();
  }
};

interface ModalProps {
  /**
   * Control if the modal is open or not.
   */
  open: boolean;
  /**
   * Is the modal closable when user press esc key.
   * Default to true.
   */
  closeOnEsc?: boolean;
  /**
   * Whether to block scrolling when dialog is open.
   */
  blockScroll?: boolean;
  /**
   * You can specify a container prop which should be of type `Element`.
   * The portal will be rendered inside that element.
   * The default behavior will create a div node and render it at the at the end of document.body.
   */
  container?: Element;
  /**
   * Animation duration in milliseconds.
   */
  animationDuration?: number;
  /**
   * Callback fired when the Modal is requested to be closed by a click on the overlay or when user press esc key.
   */
  onClose: () => void;
  /**
   * Callback fired when the escape key is pressed.
   */
  onEscKeyDown?: (event: KeyboardEvent) => void;
  /**
   * Callback fired when the Modal is open and the animation is finished.
   */
  onEntered?: () => void;
  /**
   * Callback fired when the Modal has exited and the animation is finished.
   */
  onExited?: () => void;
}

export const Modal = ({
  open,
  blockScroll = true,
  closeOnEsc = true,
  container,
  animationDuration = 500,
  onClose,
  onEscKeyDown,
  onEntered,
  onExited,
}: ModalProps) => {
  const refContainer = useRef<HTMLDivElement | null>(null);
  // Lazily create the ref instance
  // https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily
  if (refContainer.current === null && isBrowser) {
    refContainer.current = document.createElement('div');
  }

  const [showPortal, setShowPortal] = useState(open);

  const handleOpen = () => {
    modalManager.add(refContainer.current);
    if (blockScroll) {
      blockNoScroll();
    }
    if (refContainer.current && !container) {
      document.body.appendChild(refContainer.current);
    }
    document.addEventListener('keydown', handleKeydown);
  };

  const handleClose = () => {
    modalManager.remove(refContainer.current);
    if (blockScroll) {
      unblockNoScroll();
    }
    if (refContainer.current && !container) {
      document.body.removeChild(refContainer.current);
    }
    document.removeEventListener('keydown', handleKeydown);
  };

  const handleKeydown = (event: KeyboardEvent) => {
    // Only the last modal need to be escaped when pressing the esc key
    if (
      event.keyCode !== 27 ||
      !modalManager.isTopModal(refContainer.current)
    ) {
      return;
    }

    if (onEscKeyDown) {
      onEscKeyDown(event);
    }

    if (closeOnEsc) {
      onClose();
    }
  };

  useEffect(() => {
    // When the modal is rendered first time we want to block the scroll
    if (open) {
      handleOpen();
    }
    return () => {
      // When the component is unmounted directly we want to unblock the scroll
      if (showPortal) {
        handleClose();
      }
    };
  }, []);

  useEffect(() => {
    // If the open prop is changing, we need to open the modal
    if (open && !showPortal) {
      setShowPortal(true);
      handleOpen();
    } else if (!open && showPortal) {
      handleClose();
    }
  }, [open]);

  const handleEntered = () => {
    if (onEntered) {
      onEntered();
    }
  };

  const handleExited = () => {
    if (onExited) {
      onExited();
    }

    if (!open) {
      setShowPortal(false);
    }

    if (blockScroll) {
      unblockNoScroll();
    }
  };

  console.log('open', open);
  console.log('showPortal', showPortal);
  console.log('refContainer.current', refContainer.current);

  if (!showPortal) {
    return null;
  }

  return ReactDom.createPortal(
    <CSSTransition
      in={open}
      appear
      classNames={{
        appear: 'transitionEnter',
        appearActive: 'transitionEnterActive',
        enter: 'transitionEnter',
        enterActive: 'transitionEnterActive',
        exit: 'transitionExit',
        exitActive: 'transitionExitActive',
      }}
      timeout={animationDuration}
      onEntered={handleEntered}
      onExited={handleExited}
    >
      <div>Hey show me that modal</div>
    </CSSTransition>,
    container || refContainer.current!
  );
};
