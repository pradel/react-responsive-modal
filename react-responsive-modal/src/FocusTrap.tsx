import { useEffect, useRef } from 'react';
import { isBrowser } from './utils';
import {
  tabTrappingKey,
  candidateSelectors,
  getAllTabbingElements,
} from './lib/focusTrapJs';

interface FocusTrapProps {
  container?: React.RefObject<HTMLElement> | null;
}

export const FocusTrap = ({ container }: FocusTrapProps) => {
  const refLastFocus = useRef<HTMLElement | null>();
  /**
   * Handle focus lock on the modal
   */
  useEffect(() => {
    const handleKeyEvent = (event: KeyboardEvent) => {
      if (container?.current) {
        tabTrappingKey(event, container.current);
      }
    };

    if (isBrowser) {
      document.addEventListener('keydown', handleKeyEvent);
    }
    // On mount we focus on the first focusable element in the modal if there is one
    if (isBrowser && container?.current) {
      const allTabbingElements = getAllTabbingElements(container.current);
      if (allTabbingElements[0]) {
        // First we save the last focused element
        // only if it's a focusable element
        if (
          candidateSelectors.findIndex((selector) =>
            document.activeElement?.matches(selector)
          ) !== -1
        ) {
          refLastFocus.current = document.activeElement as HTMLElement;
        }
        allTabbingElements[0].focus();
      }
    }
    return () => {
      if (isBrowser) {
        document.removeEventListener('keydown', handleKeyEvent);
        // On unmount we restore the focus to the last focused element
        refLastFocus.current?.focus();
      }
    };
  }, [container]);

  return null;
};
