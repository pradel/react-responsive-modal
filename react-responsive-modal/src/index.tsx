import React, { useEffect, useState, useRef } from 'react';
import ReactDom from 'react-dom';
import cx from 'classnames';
import CloseIcon from './CloseIcon';
import { FocusTrap } from './FocusTrap';
import { modalManager, useModalManager } from './modalManager';
import { useScrollLock } from './useScrollLock';
import { isBrowser } from './utils';

const classes = {
  root: 'react-responsive-modal-root',
  overlay: 'react-responsive-modal-overlay',
  overlayAnimationIn: 'react-responsive-modal-overlay-in',
  overlayAnimationOut: 'react-responsive-modal-overlay-out',
  modalContainer: 'react-responsive-modal-container',
  modalContainerCenter: 'react-responsive-modal-containerCenter',
  modal: 'react-responsive-modal-modal',
  modalAnimationIn: 'react-responsive-modal-modal-in',
  modalAnimationOut: 'react-responsive-modal-modal-out',
  closeButton: 'react-responsive-modal-closeButton',
};

export interface ModalProps {
  /**
   * Control if the modal is open or not.
   */
  open: boolean;
  /**
   * Should the dialog be centered.
   *
   * Default to false.
   */
  center?: boolean;
  /**
   * Is the modal closable when user press esc key.
   *
   * Default to true.
   */
  closeOnEsc?: boolean;
  /**
   * Is the modal closable when user click on overlay.
   *
   * Default to true.
   */
  closeOnOverlayClick?: boolean;
  /**
   * Whether to block scrolling when dialog is open.
   *
   * Default to true.
   */
  blockScroll?: boolean;
  /**
   * Show the close icon.
   *
   * Default to true.
   */
  showCloseIcon?: boolean;
  /**
   * id attribute for the close icon button.
   */
  closeIconId?: string;
  /**
   * Custom icon to render (svg, img, etc...).
   */
  closeIcon?: React.ReactNode;
  /**
   * When the modal is open, trap focus within it.
   *
   * Default to true.
   */
  focusTrapped?: boolean;
  /**
   * You can specify a container prop which should be of type `Element`.
   * The portal will be rendered inside that element.
   * The default behavior will create a div node and render it at the at the end of document.body.
   */
  container?: Element | null;
  /**
   * An object containing classNames to style the modal.
   */
  classNames?: {
    root?: string;
    overlay?: string;
    overlayAnimationIn?: string;
    overlayAnimationOut?: string;
    modalContainer?: string;
    modal?: string;
    modalAnimationIn?: string;
    modalAnimationOut?: string;
    closeButton?: string;
    closeIcon?: string;
  };
  /**
   * An object containing the styles objects to style the modal.
   */
  styles?: {
    root?: React.CSSProperties;
    overlay?: React.CSSProperties;
    modalContainer?: React.CSSProperties;
    modal?: React.CSSProperties;
    closeButton?: React.CSSProperties;
    closeIcon?: React.CSSProperties;
  };
  /**
   * Animation duration in milliseconds.
   *
   * Default to 500.
   */
  animationDuration?: number;
  /**
   * ARIA role for modal
   *
   * Default to 'dialog'.
   */
  role?: string;
  /**
   * ARIA label for modal
   */
  ariaLabelledby?: string;
  /**
   * ARIA description for modal
   */
  ariaDescribedby?: string;
  /**
   * id attribute for modal
   */
  modalId?: string;
  /**
   * Callback fired when the Modal is requested to be closed by a click on the overlay or when user press esc key.
   */
  onClose: () => void;
  /**
   * Callback fired when the escape key is pressed.
   */
  onEscKeyDown?: (event: KeyboardEvent) => void;
  /**
   * Callback fired when the overlay is clicked.
   */
  onOverlayClick?: (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => void;
  /**
   * Callback fired when the Modal has exited and the animation is finished.
   */
  onAnimationEnd?: () => void;
  children?: React.ReactNode;
}

export const Modal = ({
  open,
  center,
  blockScroll = true,
  closeOnEsc = true,
  closeOnOverlayClick = true,
  container,
  showCloseIcon = true,
  closeIconId,
  closeIcon,
  focusTrapped = true,
  animationDuration = 300,
  classNames,
  styles,
  role = 'dialog',
  ariaDescribedby,
  ariaLabelledby,
  modalId,
  onClose,
  onEscKeyDown,
  onOverlayClick,
  onAnimationEnd,
  children,
}: ModalProps) => {
  const refModal = useRef<HTMLDivElement>(null);
  const refShouldClose = useRef<boolean | null>(null);
  const refContainer = useRef<HTMLDivElement | null>(null);
  // Lazily create the ref instance
  // https://reactjs.org/docs/hooks-faq.html#how-to-create-expensive-objects-lazily
  if (refContainer.current === null && isBrowser) {
    refContainer.current = document.createElement('div');
  }

  // The value should be false for srr, that way when the component is hydrated client side,
  // it will match the server rendered content
  const [showPortal, setShowPortal] = useState(false);

  // Hook used to manage multiple modals opened at the same time
  useModalManager(refModal, open);

  // Hook used to manage the scroll
  useScrollLock(refModal, open, showPortal, blockScroll);

  const handleOpen = () => {
    if (
      refContainer.current &&
      !container &&
      !document.body.contains(refContainer.current)
    ) {
      document.body.appendChild(refContainer.current);
    }

    document.addEventListener('keydown', handleKeydown);
  };

  const handleClose = () => {
    if (
      refContainer.current &&
      !container &&
      document.body.contains(refContainer.current)
    ) {
      document.body.removeChild(refContainer.current);
    }
    document.removeEventListener('keydown', handleKeydown);
  };

  const handleKeydown = (event: KeyboardEvent) => {
    // Only the last modal need to be escaped when pressing the esc key
    if (event.keyCode !== 27 || !modalManager.isTopModal(refModal)) {
      return;
    }

    onEscKeyDown?.(event);

    if (closeOnEsc) {
      onClose();
    }
  };

  useEffect(() => {
    return () => {
      if (showPortal) {
        // When the modal is closed or removed directly, cleanup the listeners
        handleClose();
      }
    };
  }, [showPortal]);

  useEffect(() => {
    // If the open prop is changing, we need to open the modal
    // This is also called on the first render if the open prop is true when the modal is created
    if (open && !showPortal) {
      setShowPortal(true);
      handleOpen();
    }
  }, [open]);

  const handleClickOverlay = (
    event: React.MouseEvent<HTMLDivElement, MouseEvent>
  ) => {
    if (refShouldClose.current === null) {
      refShouldClose.current = true;
    }

    if (!refShouldClose.current) {
      refShouldClose.current = null;
      return;
    }

    onOverlayClick?.(event);

    if (closeOnOverlayClick) {
      onClose();
    }

    refShouldClose.current = null;
  };

  const handleModalEvent = () => {
    refShouldClose.current = false;
  };

  const handleAnimationEnd = () => {
    if (!open) {
      setShowPortal(false);
    }

    onAnimationEnd?.();
  };

  const containerModal = container || refContainer.current;

  const overlayAnimation = open
    ? classNames?.overlayAnimationIn ?? classes.overlayAnimationIn
    : classNames?.overlayAnimationOut ?? classes.overlayAnimationOut;

  const modalAnimation = open
    ? classNames?.modalAnimationIn ?? classes.modalAnimationIn
    : classNames?.modalAnimationOut ?? classes.modalAnimationOut;

  return showPortal && containerModal
    ? ReactDom.createPortal(
        <div
          className={cx(classes.root, classNames?.root)}
          style={styles?.root}
          data-testid="root"
        >
          <div
            className={cx(classes.overlay, classNames?.overlay)}
            data-testid="overlay"
            aria-hidden={true}
            style={{
              animation: `${overlayAnimation} ${animationDuration}ms`,
              ...styles?.overlay,
            }}
          />
          <div
            ref={refModal}
            className={cx(
              classes.modalContainer,
              center && classes.modalContainerCenter,
              classNames?.modalContainer
            )}
            style={styles?.modalContainer}
            data-testid="modal-container"
            onClick={handleClickOverlay}
          >
            <div
              className={cx(classes.modal, classNames?.modal)}
              style={{
                animation: `${modalAnimation} ${animationDuration}ms`,
                ...styles?.modal,
              }}
              onMouseDown={handleModalEvent}
              onMouseUp={handleModalEvent}
              onClick={handleModalEvent}
              onAnimationEnd={handleAnimationEnd}
              id={modalId}
              role={role}
              aria-modal="true"
              aria-labelledby={ariaLabelledby}
              aria-describedby={ariaDescribedby}
              data-testid="modal"
            >
              {focusTrapped && <FocusTrap container={refModal} />}
              {children}
              {showCloseIcon && (
                <CloseIcon
                  classes={classes}
                  classNames={classNames}
                  styles={styles}
                  closeIcon={closeIcon}
                  onClick={onClose}
                  id={closeIconId}
                />
              )}
            </div>
          </div>
        </div>,
        containerModal
      )
    : null;
};

export default Modal;
