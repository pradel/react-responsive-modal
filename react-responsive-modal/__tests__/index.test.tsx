import * as React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Modal } from '../src';

describe('modal', () => {
  describe('overlay', () => {
    it('should call onClose when click on the overlay', () => {
      const onClose = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={onClose}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.click(getByTestId('modal-container'));
      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it('should disable the handler when closeOnOverlayClick is false', () => {
      const onClose = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={onClose} closeOnOverlayClick={false}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.click(getByTestId('modal-container'));
      expect(onClose).not.toHaveBeenCalled();
    });

    it('should ignore the overlay click if the event does not come from the overlay', () => {
      const onClose = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={onClose}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.click(getByTestId('modal'));
      expect(onClose).not.toHaveBeenCalled();
    });
  });

  describe('key events', () => {
    it('an invalid event should not call onClose', () => {
      const onClose = jest.fn();
      const { container } = render(
        <Modal open onClose={onClose}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.keyDown(container, { key: 'Enter', keyCode: 13 });
      expect(onClose).not.toHaveBeenCalled();
    });

    it('should not call onClose when closeOnEsc is false', () => {
      const onClose = jest.fn();
      const { container } = render(
        <Modal open onClose={onClose} closeOnEsc={false}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.keyDown(container, { keyCode: 27 });
      expect(onClose).not.toHaveBeenCalled();
    });

    it('should call onClose when pressing esc key', () => {
      const onClose = jest.fn();
      const { container } = render(
        <Modal open onClose={onClose}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.keyDown(container, { keyCode: 27 });
      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it('should call onClose of last modal only when pressing esc key when multiple modals are opened', () => {
      const onClose = jest.fn();
      const onClose2 = jest.fn();
      const { container } = render(
        <>
          <Modal open onClose={onClose}>
            <div>modal content</div>
          </Modal>
          <Modal open onClose={onClose2}>
            <div>modal content</div>
          </Modal>
        </>
      );

      fireEvent.keyDown(container, { keyCode: 27 });
      expect(onClose).not.toHaveBeenCalled();
      expect(onClose2).toHaveBeenCalledTimes(1);
    });
  });

  describe('body scroll', () => {
    it('should not block the scroll when modal is rendered closed', () => {
      render(
        <Modal open={false} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('');
    });

    it('should block the scroll when modal is rendered open', () => {
      render(
        <Modal open={true} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('hidden');
    });

    it('should block scroll when prop open change to true', () => {
      const { rerender } = render(
        <Modal open={false} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('');

      rerender(
        <Modal open={true} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('hidden');
    });

    it('should unblock scroll when prop open change to false', async () => {
      const { rerender, queryByTestId, getByTestId } = render(
        <Modal open={true} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('hidden');

      rerender(
        <Modal open={false} onClose={() => null} animationDuration={0}>
          <div>modal content</div>
        </Modal>
      );
      // Simulate the browser animation end
      fireEvent.animationEnd(getByTestId('modal'));
      await waitFor(
        () => {
          expect(queryByTestId('modal')).not.toBeInTheDocument();
        },
        { timeout: 1 }
      );

      expect(document.body.style.overflow).toBe('');
    });

    it('should unblock scroll when unmounted directly', async () => {
      const { unmount } = render(
        <Modal open={true} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('hidden');

      unmount();
      expect(document.body.style.overflow).toBe('');
    });

    it('should unblock scroll when multiple modals are opened and then closed', async () => {
      const { rerender, getAllByTestId, queryByText } = render(
        <React.Fragment>
          <Modal open onClose={() => null}>
            <div>first modal</div>
          </Modal>
          <Modal open onClose={() => null}>
            <div>second modal</div>
          </Modal>
        </React.Fragment>
      );
      expect(document.body.style.overflow).toBe('hidden');

      // We close one modal, the scroll should be locked
      rerender(
        <React.Fragment>
          <Modal open onClose={() => null}>
            <div>first modal</div>
          </Modal>
          <Modal open={false} onClose={() => null}>
            <div>second modal</div>
          </Modal>
        </React.Fragment>
      );

      fireEvent.animationEnd(getAllByTestId('modal')[1]);
      await waitFor(
        () => {
          expect(queryByText(/second modal/)).not.toBeInTheDocument();
        },
        { timeout: 1 }
      );
      expect(document.body.style.overflow).toBe('hidden');

      // We close the second modal, the scroll should be unlocked
      rerender(
        <React.Fragment>
          <Modal open={false} onClose={() => null}>
            <div>first modal</div>
          </Modal>
          <Modal open={false} onClose={() => null}>
            <div>second modal</div>
          </Modal>
        </React.Fragment>
      );

      fireEvent.animationEnd(getAllByTestId('modal')[0]);
      await waitFor(
        () => {
          expect(queryByText(/first modal/)).not.toBeInTheDocument();
        },
        { timeout: 1 }
      );
      expect(document.body.style.overflow).toBe('');
    });

    it('should unblock scroll when one modal is closed and the one still open has blockScroll set to false', async () => {
      const { rerender, getAllByTestId, queryByText } = render(
        <React.Fragment>
          <Modal open blockScroll={false} onClose={() => null}>
            <div>first modal</div>
          </Modal>
          <Modal open onClose={() => null}>
            <div>second modal</div>
          </Modal>
        </React.Fragment>
      );
      expect(document.body.style.overflow).toBe('hidden');

      // We close one modal, the scroll should be unlocked as remaining modal is not locking the scroll
      rerender(
        <React.Fragment>
          <Modal open blockScroll={false} onClose={() => null}>
            <div>first modal</div>
          </Modal>
          <Modal open={false} onClose={() => null}>
            <div>second modal</div>
          </Modal>
        </React.Fragment>
      );

      fireEvent.animationEnd(getAllByTestId('modal')[1]);
      await waitFor(
        () => {
          expect(queryByText(/second modal/)).not.toBeInTheDocument();
        },
        { timeout: 1 }
      );
      expect(document.body.style.overflow).toBe('');
    });
    it('should reserve scroll bar gap', () => {
      const scrollBarWidth = 42;
      const innerWidth = 500;
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: innerWidth,
      });
      Object.defineProperty(document.documentElement, 'clientWidth', {
        writable: true,
        configurable: true,
        value: innerWidth - scrollBarWidth,
      });
      render(
        <Modal open={true} onClose={() => null} reserveScrollBarGap={true}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.paddingRight).toBe(`${scrollBarWidth}px`);
    });
  });

  describe('closeIcon', () => {
    it('should render the closeIcon by default', () => {
      const { getByTestId } = render(
        <Modal open onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );

      expect(getByTestId('close-button')).toMatchSnapshot();
    });

    it('should hide closeIcon when showCloseIcon is false', () => {
      const { queryByTestId } = render(
        <Modal open onClose={() => null} showCloseIcon={false}>
          <div>modal content</div>
        </Modal>
      );

      expect(queryByTestId('close-button')).toBeNull();
    });

    it('should call onClose when clicking on the icon', () => {
      const onClose = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={onClose}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.click(getByTestId('close-button'));
      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe('render', () => {
    it('should render null when then modal is not open', () => {
      const { queryByText } = render(
        <Modal open={false} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(queryByText(/modal content/)).toBeNull();
    });

    it('should render the content when modal is open', () => {
      const { queryByText } = render(
        <Modal open onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(queryByText(/modal content/)).toBeTruthy();
    });
  });

  describe('lifecycle', () => {
    it('should show modal when prop open change to true', () => {
      const { queryByTestId, rerender } = render(
        <Modal open={false} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(queryByTestId('modal')).toBeNull();
      rerender(
        <Modal open={true} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(queryByTestId('modal')).toBeTruthy();
    });

    it('should hide modal when prop open change to false', async () => {
      const { getByTestId, queryByTestId, rerender } = render(
        <Modal open={true} onClose={() => null} animationDuration={0.01}>
          <div>modal content</div>
        </Modal>
      );
      expect(queryByTestId('modal')).toBeTruthy();
      rerender(
        <Modal open={false} onClose={() => null} animationDuration={0.01}>
          <div>modal content</div>
        </Modal>
      );
      fireEvent.animationEnd(getByTestId('modal'));
      expect(queryByTestId('modal')).toBeNull();
    });
  });

  describe('prop: center', () => {
    it('should not apply center class by default', async () => {
      const { getByTestId } = render(
        <Modal open onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );

      expect(getByTestId('modal-container').classList.length).toBe(1);
      expect(
        getByTestId('modal-container').classList.contains(
          'react-responsive-modal-containerCenter'
        )
      ).toBeFalsy();
    });

    it('should apply center class to modal', async () => {
      const { getByTestId } = render(
        <Modal open onClose={() => null} center>
          <div>modal content</div>
        </Modal>
      );

      expect(getByTestId('modal-container').classList.length).toBe(2);
      expect(
        getByTestId('modal-container').classList.contains(
          'react-responsive-modal-containerCenter'
        )
      ).toBeTruthy();
    });
  });

  describe('prop: closeIcon', () => {
    it('should render custom icon instead of the default one', async () => {
      const { queryByTestId, getByTestId } = render(
        <Modal
          open
          onClose={() => null}
          closeIcon={<div data-testid="custom-icon">custom icon</div>}
        >
          <div>modal content</div>
        </Modal>
      );

      expect(queryByTestId('close-icon')).toBeNull();
      expect(getByTestId('custom-icon')).toMatchSnapshot();
    });
  });

  describe('prop: classNames', () => {
    it('should apply custom classes to the modal', async () => {
      const { getByTestId } = render(
        <Modal
          open
          onClose={() => null}
          classNames={{
            overlay: 'custom-overlay',
            modal: 'custom-modal',
            closeButton: 'custom-closeButton',
            closeIcon: 'custom-closeIcon',
          }}
        >
          <div>modal content</div>
        </Modal>
      );

      expect(
        getByTestId('overlay').classList.contains('custom-overlay')
      ).toBeTruthy();
      expect(
        getByTestId('modal').classList.contains('custom-modal')
      ).toBeTruthy();
      expect(
        getByTestId('close-button').classList.contains('custom-closeButton')
      ).toBeTruthy();
      expect(
        getByTestId('close-icon').classList.contains('custom-closeIcon')
      ).toBeTruthy();
    });
  });

  describe('prop: blockScroll', () => {
    it('should not block the scroll when modal is opened and blockScroll is false', () => {
      render(
        <Modal open blockScroll={false} onClose={() => null}>
          <div>modal content</div>
        </Modal>
      );
      expect(document.body.style.overflow).toBe('');
    });
  });

  describe('prop: onEscKeyDown', () => {
    it('should be called when esc key is pressed', async () => {
      const onEscKeyDown = jest.fn();
      const { container } = render(
        <Modal open onClose={() => null} onEscKeyDown={onEscKeyDown}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.keyDown(container, { keyCode: 27 });
      expect(onEscKeyDown).toHaveBeenCalledTimes(1);
    });
  });

  describe('prop: onOverlayClick', () => {
    it('should be called when user click on overlay', async () => {
      const onOverlayClick = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={() => null} onOverlayClick={onOverlayClick}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.click(getByTestId('modal-container'));
      expect(onOverlayClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('prop: onAnimationEnd', () => {
    it('should be called when the animation is finished', async () => {
      const onAnimationEnd = jest.fn();
      const { getByTestId } = render(
        <Modal open onClose={() => null} onAnimationEnd={onAnimationEnd}>
          <div>modal content</div>
        </Modal>
      );

      fireEvent.animationEnd(getByTestId('modal'));
      expect(onAnimationEnd).toHaveBeenCalledTimes(1);
    });
  });

  describe('prop: containerId', () => {
    it('should renders container div with id', async () => {
      const containerId = 'container-id';
      const { getByTestId } = render(
        <Modal open onClose={() => null} containerId={containerId}>
          <div>modal content</div>
        </Modal>
      );

      const containerModal = getByTestId('modal-container');
      expect(containerModal.getAttribute('id')).toBe(containerId);
      expect(document.getElementById(containerId)).toBeInTheDocument();
    });
  });

  describe('prop: modalId', () => {
    it('should renders modal div with id', async () => {
      const modalId = 'modal-id';
      const { getByTestId } = render(
        <Modal open onClose={() => null} modalId={modalId}>
          <div>modal content</div>
        </Modal>
      );

      const modal = getByTestId('modal');
      expect(modal.getAttribute('id')).toBe(modalId);
      expect(document.getElementById(modalId)).toBeInTheDocument();
    });
  });
});
