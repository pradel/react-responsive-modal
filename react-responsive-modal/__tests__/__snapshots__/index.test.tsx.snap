// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal closeIcon should render the closeIcon by default 1`] = `
<button
  class="react-responsive-modal-closeButton"
  data-testid="close-button"
>
  <svg
    data-testid="close-icon"
    height="28"
    viewBox="0 0 36 36"
    width="28"
  >
    <path
      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
    />
  </svg>
</button>
`;

exports[`modal prop: closeIcon should render custom icon instead of the default one 1`] = `
<div
  data-testid="custom-icon"
>
  custom icon
</div>
`;
